peer/brontide_test.go

```diff
@@ -1118,3 +1118,136 @@ func TestPeerCustomMessage(t *testing.T) {
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			t.Parallel()
+
+			dbAlice, err := channeldb.Open(t.TempDir())
+			require.NoError(t, err)
+
+			aliceKey, err := btcec.NewPrivateKey()
+			require.NoError(t, err)
+
+			writeBufferPool := pool.NewWriteBuffer(
+				pool.DefaultWriteBufferGCInterval,
+				pool.DefaultWriteBufferExpiryInterval,
+			)
+
+			writePool := pool.NewWrite(
+				writeBufferPool, 1, timeout,
+			)
+			require.NoError(t, writePool.Start())
+
+			readBufferPool := pool.NewReadBuffer(
+				pool.DefaultReadBufferGCInterval,
+				pool.DefaultReadBufferExpiryInterval,
+			)
+
+			readPool := pool.NewRead(
+				readBufferPool, 1, timeout,
+			)
+			require.NoError(t, readPool.Start())
+
+			mockConn := newMockConn(t, 1)
+
+			receivedCustomChan := make(chan *customMsg)
+
+			remoteKey := [33]byte{8}
+
+			notifier := &mock.ChainNotifier{
+				SpendChan: make(chan *chainntnfs.SpendDetail),
+				EpochChan: make(chan *chainntnfs.BlockEpoch),
+				ConfChan:  make(chan *chainntnfs.TxConfirmation),
+			}
+
+			channelNotifier := channelnotifier.New(dbAlice.ChannelStateDB())
+			require.NoError(t, channelNotifier.Start())
+			t.Cleanup(func() {
+				require.NoError(t, channelNotifier.Stop(),
+					"stop channel notifier failed")
+			})
+
+			alicePeer := NewBrontide(Config{
+				PubKeyBytes: remoteKey,
+				ChannelDB:   dbAlice.ChannelStateDB(),
+				Addr: &lnwire.NetAddress{
+					IdentityKey: aliceKey.PubKey(),
+				},
+				PrunePersistentPeerConnection: func([33]byte) {},
+				Features:                      lnwire.EmptyFeatureVector(),
+				LegacyFeatures:                lnwire.EmptyFeatureVector(),
+				WritePool:                     writePool,
+				ReadPool:                      readPool,
+				Conn:                          mockConn,
+				ChainNotifier:                 notifier,
+				HandleCustomMessage: func(
+					peer [33]byte, msg *lnwire.Custom) error {
+
+					receivedCustomChan <- &customMsg{
+						peer: peer,
+						msg:  *msg,
+					}
+					return nil
+				},
+				PongBuf:         make([]byte, lnwire.MaxPongBytes),
+				ChannelNotifier: channelNotifier,
+			})
+
+			// Set up the init sequence.
+			go func() {
+				// Read init message.
+				<-mockConn.writtenMessages
+
+				// Write the init reply message.
+				initReplyMsg := lnwire.NewInitMessage(
+					lnwire.NewRawFeatureVector(
+						lnwire.DataLossProtectRequired,
+					),
+					lnwire.NewRawFeatureVector(),
+				)
+				var b bytes.Buffer
+				_, err = lnwire.WriteMessage(&b, initReplyMsg, 0)
+				require.NoError(t, err)
+
+				mockConn.readMessages <- b.Bytes()
+			}()
+			
+			ready := make(chan struct{})
+
+			// Start the peer.
+			require.NoError(t, alicePeer.Start())
+
+		    close(ready)
+
+			receivedData := append([]byte{byte(testCase.messageType >> 8), byte(testCase.messageType)}, testCase.messageData...)
+			mockConn.readMessages <- receivedData
+
+			alicePeer.WaitForDisconnect(ready)
```

Comment:
The way `WaitForDisconnect` is used here confuses me.

- Why are we creating and closing the `ready` channel beforehand?
- Does `WaitForDisconnect` hang if we send an odd message and no disconnect happens?

Reply:
Yes sorry there were some oversights. There was some unexpected behaviour due to t.Parallel(), so that's been removed and now TestPeerUnknownMessage is running as expected once again.

> Why are we creating and closing the ready channel beforehand?

It's to ensure that the peer is fully started and ready to receive messages before sending any messages to it

> Does WaitForDisconnect hang if we send an odd message and no disconnect happens?

Yes, that's the intended purpose.. the hanging wasn't taking place as expected due to t.Parallel(). Now it is. I've just added a simple conditional:

```golang
    if (testCase.shouldClose){
        alicePeer.WaitForDisconnect(ready)
    } else {
        time.Sleep(100 * time.Millisecond)
    }
```
as this is nice and simple on the eyes. But to properly test the hanging case, i could add a `select` or a `timer`. Please advise.

Thank you for this review, it's much appreciated.

Reply:
> Yes sorry there were some oversights. There was some unexpected behaviour due to t.Parallel(), so that's been removed and now TestPeerUnknownMessage is running as expected once again.

Why doesn't parallel execution work?  Ideally none of the test cases should affect each other.

> 
> > Why are we creating and closing the ready channel beforehand?
> 
> It's to ensure that the peer is fully started and ready to receive messages before sending any messages to it

The `ready` channel isn't passed to the peer, so there doesn't seem to be a reason to have the channel at all, except to satisfy the function parameters.  And in that case, it would be clearer to put the channel creation and closure directly before the `WaitForDisconnect` (or maybe use a `nil` channel).

> 
> > Does WaitForDisconnect hang if we send an odd message and no disconnect happens?
> 
> Yes, that's the intended purpose.. the hanging wasn't taking place as expected due to t.Parallel(). Now it is. I've just added a simple conditional:
> 
> ```go
>     if (testCase.shouldClose){
>         alicePeer.WaitForDisconnect(ready)
>     } else {
>         time.Sleep(100 * time.Millisecond)
>     }
> ```
> 
> as this is nice and simple on the eyes. But to properly test the hanging case, i could add a `select` or a `timer`. Please advise.

I think a select with a timeout is what we want here.  Something like:

```go
func peerDisconnected(p *Brontide) bool {
  disconnected := make(chan struct{}, 1)
  go func() {
    alicePeer.WaitForDisconnect(nil)
    close(disconnected)
  }()

  select {
  case <-disconnected:
    return true
  case <-time.After(timeout):
    return false
  }
}

...

require.Equal(testcase.shouldClose, peerDisconnected(alicePeer))
```